File: 2930-number-of-strings-which-can-be-rearranged-to-contain-substring/2930-number-of-strings-which-can-be-rearranged-to-contain-substring.cpp
using ll = long long;
using ti3 = tuple<ll,ll,ll>;
using vi = vector<ll>;
using vvi = vector<vi>;
const ll mod = 1e9+7;

class Solution {
public:
    vvi mat_mul(vvi &a, vvi &b) {
        int n = a.size();
        vvi ret(n, vi(n));
        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                for (int k=0; k<n; k++) {
                    ret[i][j] += a[i][k] * b[k][j];
                    ret[i][j] %= mod;
                }
            }
        }
        return ret;
    }
    
    vvi mat_exp(vvi a, int e) {
        int n = a.size();
        vvi ret(n, vi(n));
        for (int i=0; i<n; i++) ret[i][i] = 1;
        while (e) {
            if (e&1) ret = mat_mul(ret, a);
            e >>= 1;
            a = mat_mul(a,a);
        }
        return ret;
    }

    int stringCount(int n) {
        map<ti3,int> mp;
        ll m = 0;
        for (int i=0; i<2; i++) {
            for (int j=0; j<2; j++) {
                for (int k=0; k<3; k++) {
                    mp[{i,j,k}] = m++;
                }
            }
        }
        
        vvi a(m, vi(m));
        for (int i=0; i<2; i++) {
            for (int j=0; j<2; j++) {
                for (int k=0; k<3; k++) {
                    int ni = min(1,i+1);
                    int nj = min(1,j+1);
                    int nk = min(2,k+1);
                    ll u = mp[{i,j,k}];
                    a[u][u] += 23;
                    a[u][mp[{ni,j,k}]] += 1;
                    a[u][mp[{i,nj,k}]] += 1;
                    a[u][mp[{i,j,nk}]] += 1;
                }
            }
        }
        
        a = mat_exp(a, n);
        return a[0][m-1];        
    }
};

File: 2930-number-of-strings-which-can-be-rearranged-to-contain-substring/2930-number-of-strings-which-can-be-rearranged-to-contain-substring.kt
typealias ll = Long
typealias ti3 = Triple<Int,Int,Int>
typealias vi = List<Long>
typealias vvi = List<vi>
const val mod = 1e9.toLong() + 7

    
class Solution {
    private fun matMul(a: vvi, b: vvi): vvi {
        val n = a.size
        return (0 until n).map { i ->
            (0 until n).map { j ->
                (0 until n).sumOf { k ->
                    a[i][k]*b[k][j]
                }.rem(mod)
            }
        }
    }

    private fun matExp(a: vvi, e: Int): vvi {
        val n = a.size
        val ret = (0 until n).map { i ->
            (0 until n).map { j ->
                if (i == j) 1L else 0L
            }
        }.toList()

        tailrec fun matExpHelper(exp: Int, base: vvi, res: vvi): vvi {
            return when {
                exp == 0 -> res
                exp and 1 == 1 -> matExpHelper(exp shr 1, matMul(base, base), matMul(res, base))
                else -> matExpHelper(exp shr 1, matMul(base, base), res)
            }
        }

        return matExpHelper(e, a, ret)
    }


    fun stringCount(n: Int): Int {
        val mp = (0 until 2).flatMap { i ->
            (0 until 2).flatMap { j ->
                (0 until 3).map { k ->
                    Triple(i, j, k)
                }
            }
        }.withIndex().associate { (index, triple) ->
            triple to index
        }.toMap()
        
        val m = mp.size
        val a = MutableList(m) { MutableList(m) { 0L } }
        
        (0 until 2).forEach { i ->
            (0 until 2).forEach { j ->
                (0 until 3).forEach { k ->
                    val ni = minOf(1, i + 1)
                    val nj = minOf(1, j + 1)
                    val nk = minOf(2, k + 1)
                    val u = mp[Triple(i, j, k)]!!
                    val v1 = mp[Triple(ni, j, k)]!!
                    val v2 = mp[Triple(i, nj, k)]!!
                    val v3 = mp[Triple(i, j, nk)]!!
                    a[u][u]+= 23L
                    a[u][v1]+= 1L
                    a[u][v2]+= 1L
                    a[u][v3]+= 1L
                }
            }
        }
        
        return matExp(a, n)[0][m-1].toInt()
    }
}  